odoo/models.py

```diff
@@ -2437,27 +2438,70 @@ def _inherits_join_add(self, current_model, parent_model_name, query):
+        :param allow_related: whether we make joins for explicit
+                              (i.e. not inherited) related non-stored field
         :return: qualified name of field, to be used in SELECT clause
         """
         # INVARIANT: alias is the SQL alias of model._table in query
         model, field = self, self._fields[fname]
-        while field.inherited:
-            # retrieve the parent model where field is inherited from
-            parent_model = self.env[field.related_field.model_name]
-            parent_fname = field.related[0]
-            # JOIN parent_model._table AS parent_alias ON alias.parent_fname = parent_alias.id
-            parent_alias = query.left_join(
-                alias, parent_fname, parent_model._table, 'id', parent_fname,
-            )
-            model, alias, field = parent_model, parent_alias, field.related_field
+        def process_inherited(model, alias, field, check_auto_join=False):
+            while field.inherited:
+                # retrieve the parent model where field is inherited from
+                parent_model = self.env[field.related_field.model_name]
+                parent_fname = field.related[0]
+                if check_auto_join:
+                    join_field = model._fields[parent_fname]
+                    if not join_field.auto_join:
+                        raise AccessError(_("Field %s has %s and can't be used without sudo mode") % (
+                            join_field, 'auto_join=False'
+                        ))
+                # JOIN parent_model._table AS parent_alias ON alias.parent_fname = parent_alias.id
+                parent_alias = query.left_join(
+                    alias, parent_fname, parent_model._table, 'id', parent_fname,
+                )
+                model, alias, field = parent_model, parent_alias, field.related_field
+            return model, alias, field
+
+
+        model, alias, field = process_inherited(model, alias, field)
+        def process_related(model, alias, field, sudo):
+            sudo = sudo or field.compute_sudo
+            related_field = field.related_field
+            main_field = field
+            for join_fname in field.related[:-1]:
+                field = model._fields[join_fname]
+                if field.inherited:
+                    model, alias, field = process_inherited(model, alias, field, check_auto_join=not sudo)
+                if field.related and not field.store:
+                    if not (sudo or field.auto_join):
+                        # TODO: user may have access to field.model_name and should not get this access error message
+                        raise AccessError(_("Non-stored related field %s with %s is passed via field %s with %s. You can use this field in sudo mode only") % (
+                            main_field, 'compute_sudo=False',
+                            field, 'auto_join=False'
+                        ))
```

Comment:
@rco-odoo What do you think about this TODO? If it's may be useful, shall I implement it in this PR or leave for now?